### ISSUE TITLE
Remove `Event.INIT`; move `Event.TRAINING_START` to before the DDP

### ISSUE THREAD

**Issue by ravi-mosaicml** [mosaicml/composer#157]
Once #43 is merged in, then `Event.INIT` can be removed.

`Event.TRAINING_START` can be moved up to the line right before `ddp.prepare_module()`. All existing algorithms that currently run on training start would be compatible with this change.

* Augmix: :heavy_check_mark: 
* Colout: :heavy_check_mark: 
* Curriculum Learning: @moinnadeem what is being done on training start? I do not see any inputs or outputs being preserved.
* Randaugment: :heavy_check_mark: 
* Sam: :heavy_check_mark:. In fact, it might be broken as it currently runs on training_start, after optimizers have been moved onto the device
* Scale Schedule: :heavy_check_mark: 
* Seq Length Warmup: @moinnadeem what is being done on training start? I do not see any inputs or outputs being preserved.
* SWA: :heavy_check_mark: This change would simplify this algorithm, as it would apply before the model is moved onto the device (and it wouldn't need to take that into account anymore)

TODO:
- [x] Implement and merge #43 
- [x] Fix algorithms as described above
- [x] Refactor the training loop to support this change.
- [x] Update documentation. We frequently refer to `TRAINING_START` happening after the DDP fork and all DDP madness. This changes that. Will need to take a throughout read through the `docs/` folder.

**Comment by jbloxham** [comment]
I'm not sure how I feel about removing `INIT`. I definitely agree it's hacky how `Event.INIT` currently needs to exist between model initialization and optimizer initialization to accommodate surgery, but #43 will allow us to fix those issues without just removing the entire event. I think `INIT` should be called at the end of `Trainer` initialization, and `TRAINING_START` should be called after `Trainer.fit()` is called. Note also that `prepare_ddp_module()` is now called within `Trainer.__init__()`, not `Trainer.fit()`

**Comment by ravi-mosaicml** [comment]
Surgery is fixed in #249, so we can now address this issue.

**Comment by ravi-mosaicml** [comment]
Addressed in #263. Kept Event.INIT; renamed Event.TRAINING_START to Event.FIT_START
